In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, roc_auc_score

In [2]:
users = pd.read_csv('takehome_users.csv', encoding='latin-1')
engagement = pd.read_csv('takehome_user_engagement.csv')

users.head()

,object_id,creation_time,name,email,creation_source,last_session_creation_time,opted_in_to_mailing_list,enabled_for_marketing_drip,org_id,invited_by_user_id
0,1,2014-04-22 03:53:30,Clausen August,AugustCClausen@yahoo.com,GUEST_INVITE,1.398139e+09,1,0,11,10803.0
1,2,2013-11-15 03:45:04,Poole Matthew,MatthewPoole@gustr.com,ORG_INVITE,1.396238e+09,0,0,1,316.0
2,3,2013-03-19 23:14:52,Bottrill Mitchell,MitchellBottrill@gustr.com,ORG_INVITE,1.363735e+09,0,0,94,1525.0
3,4,2013-05-21 08:09:28,Clausen Nicklas,NicklasSClausen@yahoo.com,GUEST_INVITE,1.369210e+09,0,0,1,5151.0
4,5,2013-01-17 10:14:20,Raw Grace,GraceRaw@yahoo.com,GUEST_INVITE,1.358850e+09,0,0,193,5240.0


In [3]:
engagement.head()

,time_stamp,user_id,visited
0,2014-04-22 03:53:30,1,1
1,2013-11-15 03:45:04,2,1
2,2013-11-29 03:45:04,2,1
3,2013-12-09 03:45:04,2,1
4,2013-12-25 03:45:04,2,1


In [4]:
#create target variable "adopted" 
engagement["time_stamp"] = pd.to_datetime(engagement["time_stamp"])
engagement["date"] = engagement["time_stamp"].dt.date

# one row per user per login day
daily_logins = (
    engagement[["user_id", "date"]]
    .drop_duplicates()
    .sort_values(["user_id", "date"])
)

def is_adopted(dates):
    dates = pd.to_datetime(dates).sort_values().reset_index(drop=True)
    for i in range(len(dates) - 2):
        if (dates.iloc[i + 2] - dates.iloc[i]).days <= 6:
            return 1
    return 0

adopted = (
    daily_logins.groupby("user_id")["date"]
    .apply(is_adopted)
    .reset_index(name="adopted_user")
)

In [5]:
adopted.head()

,user_id,adopted_user
0,1,0
1,2,1
2,3,0
3,4,0
4,5,0


In [6]:
#merge adopted with users into one df
users = users.rename(columns={"object_id": "user_id"})
df = users.merge(adopted, on="user_id", how="left")
df["adopted_user"] = df["adopted_user"].fillna(0).astype(int)

df.head()

,user_id,creation_time,name,email,creation_source,last_session_creation_time,opted_in_to_mailing_list,enabled_for_marketing_drip,org_id,invited_by_user_id,adopted_user
0,1,2014-04-22 03:53:30,Clausen August,AugustCClausen@yahoo.com,GUEST_INVITE,1.398139e+09,1,0,11,10803.0,0
1,2,2013-11-15 03:45:04,Poole Matthew,MatthewPoole@gustr.com,ORG_INVITE,1.396238e+09,0,0,1,316.0,1
2,3,2013-03-19 23:14:52,Bottrill Mitchell,MitchellBottrill@gustr.com,ORG_INVITE,1.363735e+09,0,0,94,1525.0,0
3,4,2013-05-21 08:09:28,Clausen Nicklas,NicklasSClausen@yahoo.com,GUEST_INVITE,1.369210e+09,0,0,1,5151.0,0
4,5,2013-01-17 10:14:20,Raw Grace,GraceRaw@yahoo.com,GUEST_INVITE,1.358850e+09,0,0,193,5240.0,0


In [7]:
#feature engineering
df["creation_time"] = pd.to_datetime(df["creation_time"])
df["last_session_creation_time"] = pd.to_datetime(
    df["last_session_creation_time"], unit="s"
)

df["has_inviter"] = df["invited_by_user_id"].notna().astype(int)
df["email_domain"] = df["email"].str.split("@").str[-1]

# Keep only common domains, group the rest
top_domains = df["email_domain"].value_counts().head(10).index
df["email_domain_grouped"] = np.where(
    df["email_domain"].isin(top_domains),
    df["email_domain"],
    "other"
)

df["account_age_days"] = (
    df["last_session_creation_time"].max() - df["creation_time"]
).dt.days

# Early engagement feature: number of logins in first 7 days after signup
engagement = engagement.merge(
    df[["user_id", "creation_time"]],
    on="user_id",
    how="left"
)

engagement["days_after_signup"] = (
    engagement["time_stamp"] - engagement["creation_time"]
).dt.days

first_week_logins = (
    engagement[engagement["days_after_signup"].between(0, 7)]
    .groupby("user_id")
    .size()
    .reset_index(name="first_week_logins")
)

df = df.merge(first_week_logins, on="user_id", how="left")
df["first_week_logins"] = df["first_week_logins"].fillna(0)

df.head()

,user_id,creation_time,name,email,creation_source,last_session_creation_time,opted_in_to_mailing_list,enabled_for_marketing_drip,org_id,invited_by_user_id,adopted_user,has_inviter,email_domain,email_domain_grouped,account_age_days,first_week_logins
0,1,2014-04-22 03:53:30,Clausen August,AugustCClausen@yahoo.com,GUEST_INVITE,2014-04-22 03:53:30,1,0,11,10803.0,0,1,yahoo.com,yahoo.com,45,1.0
1,2,2013-11-15 03:45:04,Poole Matthew,MatthewPoole@gustr.com,ORG_INVITE,2014-03-31 03:45:04,0,0,1,316.0,1,1,gustr.com,gustr.com,203,1.0
2,3,2013-03-19 23:14:52,Bottrill Mitchell,MitchellBottrill@gustr.com,ORG_INVITE,2013-03-19 23:14:52,0,0,94,1525.0,0,1,gustr.com,gustr.com,443,1.0
3,4,2013-05-21 08:09:28,Clausen Nicklas,NicklasSClausen@yahoo.com,GUEST_INVITE,2013-05-22 08:09:28,0,0,1,5151.0,0,1,yahoo.com,yahoo.com,381,1.0
4,5,2013-01-17 10:14:20,Raw Grace,GraceRaw@yahoo.com,GUEST_INVITE,2013-01-22 10:14:20,0,0,193,5240.0,0,1,yahoo.com,yahoo.com,505,1.0


In [8]:
features = [
    "creation_source",
    "opted_in_to_mailing_list",
    "enabled_for_marketing_drip",
    "org_id",
    "has_inviter",
    "email_domain_grouped",
    "account_age_days",
    "first_week_logins"
]

X = df[features].copy()
y = df["adopted_user"]

categorical_cols = ["creation_source", "org_id", "email_domain_grouped"]
numeric_cols = [
    "opted_in_to_mailing_list",
    "enabled_for_marketing_drip",
    "has_inviter",
    "account_age_days",
    "first_week_logins"
]

preprocessor = ColumnTransformer([
    ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_cols),
    ("num", StandardScaler(), numeric_cols)
])

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.25,
    random_state=42,
    stratify=y
)

In [10]:
# Logistic Regression Baseline

log_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", LogisticRegression(max_iter=1000))
])

log_model.fit(X_train, y_train)

log_preds = log_model.predict(X_test)
log_probs = log_model.predict_proba(X_test)[:, 1]

print(classification_report(y_test, log_preds))
print("ROC-AUC:", roc_auc_score(y_test, log_probs))

              precision    recall  f1-score   support

           0       0.91      0.98      0.94      2600
           1       0.74      0.37      0.50       400

    accuracy                           0.90      3000
   macro avg       0.82      0.68      0.72      3000
weighted avg       0.89      0.90      0.88      3000

ROC-AUC: 0.8173173076923077


In [11]:
# Random Forest

rf_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", RandomForestClassifier(
        n_estimators=300,
        random_state=42,
        class_weight="balanced"
    ))
])

rf_model.fit(X_train, y_train)

rf_preds = rf_model.predict(X_test)
rf_probs = rf_model.predict_proba(X_test)[:, 1]

print(classification_report(y_test, rf_preds))
print("ROC-AUC:", roc_auc_score(y_test, rf_probs))

              precision    recall  f1-score   support

           0       0.90      0.97      0.94      2600
           1       0.67      0.33      0.44       400

    accuracy                           0.89      3000
   macro avg       0.79      0.65      0.69      3000
weighted avg       0.87      0.89      0.87      3000

ROC-AUC: 0.8164326923076923


Logistic Regression model performed better, with a slightly higher recall, f1-score, and precision. It is also more interpretable. We will choose this model.

In [12]:
# Feature importance / coefficients from logistic regression

feature_names = log_model.named_steps["preprocessor"].get_feature_names_out()
coefficients = log_model.named_steps["model"].coef_[0]

coef_df = pd.DataFrame({
    "feature": feature_names,
    "coefficient": coefficients
})

coef_df["abs_coefficient"] = coef_df["coefficient"].abs()

coef_df.sort_values("abs_coefficient", ascending=False).head(15)

,feature,coefficient,abs_coefficient
435,num__first_week_logins,1.572279,1.572279
240,cat__org_id_235,1.480107,1.480107
122,cat__org_id_117,1.431807,1.431807
27,cat__org_id_22,-1.281297,1.281297
399,cat__org_id_394,-1.266004,1.266004
392,cat__org_id_387,1.189428,1.189428
5,cat__org_id_0,-1.132144,1.132144
416,cat__org_id_411,-1.110688,1.110688
91,cat__org_id_86,-1.091012,1.091012
190,cat__org_id_185,1.088608,1.088608


## Final Findings

An adopted user was defined as a user who logged into the product on at least three separate days within a seven-day period.

I initially trained logistic regression and random forest classification models using only static account metadata such as signup source, organization membership, mailing preferences, and inviter information. These initial models performed poorly, with ROC-AUC scores near 0.60, suggesting that signup metadata alone was not sufficient to predict future adoption.

To improve performance, I engineered an additional behavioral feature: the number of logins during a user's first week after signup. Adding this feature substantially improved predictive performance, increasing ROC-AUC to approximately 0.82 for both models.

After feature engineering, logistic regression slightly outperformed random forest with higher recall, precision, F1-score, and ROC-AUC for adopted users. Because it also provides interpretable coefficients, logistic regression was selected as the final model.

The strongest predictor by far was first-week login activity, indicating that early engagement behavior is highly associated with long-term adoption. Several organization-related features also appeared important, suggesting that adoption behavior may vary across organizations or teams.

Overall, the analysis suggests that early user engagement is much more predictive of future adoption than static signup metadata alone.